[DAG] Replace getValid*ShiftAmountConstant helpers with getValid*ShiftAmount helpers to support KnownBits analysis

RKSimon opened the pull request:
The getValidShiftAmountConstant/getValidMinimumShiftAmountConstant/getValidMaximumShiftAmountConstant helpers only worked with constant shift amounts, which could be problematic after type legalization (e.g. v2i64 might be split into v4i32 on some targets such as 32-bit x86, Thumb2 MVE).

This patch proposes we generalize these helpers to work with KnownBits if a scalar/buildvector constant isn't available.

Most restrictions are the same - the helper fails if any shift amount is out of bounds, getValidShiftConstant must be a specific constant uniform etc.

However, getValidMinimumShiftAmount/getValidMaximumShiftAmount now can return bounds values that aren't values in the actual data, as they are based off the common KnownBits of every vector element.

This addresses feedback on #92096

RKSimon committed:
[DAG] Replace getValid*ShiftAmountConstant helpers with getValid*ShiftAmount helpers to support KnownBits analysis

The getValidShiftAmountConstant/getValidMinimumShiftAmountConstant/getValidMaximumShiftAmountConstant helpers only worked with constant shift amounts, which could be problematic after type legalization (e.g. v2i64 might be split into v4i32 on some targets such as Thumb2 MVE).

This patch proposes we generalize these helpers to work with KnownBits if a scalar/buildvector constant isn't available.

Most restrictions are the same - the helper fails if any shift amount is out of bounds, getValidShiftConstant must be a specific constant uniform etc.

However, getValidMinimumShiftAmount/getValidMaximumShiftAmount now can return bounds values that aren't values in the actual data, as they are based off the common KnownBits of every vector element.

This addresses feedback on #92096

goldsteinn commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -3020,91 +3020,111 @@ SelectionDAG::getValidShiftAmountConstant(SDValue V,
+  } else {
+    KnownBits KnownAmt =
+        computeKnownBits(V.getOperand(1), DemandedElts, Depth + 1);
+    if (KnownAmt.isConstant() && KnownAmt.getConstant().ult(BitWidth))
```
IMO if its constant only, Depth should start at MaxDepth - 2, otherwise think there is a lot of wasted compute.

RKSimon replied:
The problem we're hitting is that in many cases its not just a bitcast of a constant, its often more complex than that and when trying to use value tracking during legalization we only get one chance, we can't wait for combines to simplify things.

goldsteinn replied:
Okay, fair enough (maybe mention in a comment? This seems a bit non-intuitive).

RKSimon commented:
https://llvm-compile-time-tracker.com/compare.php?from=84729c9df30b29d5f4e903ad71235a6aa0c764d6&to=d8ee31077ae6f1acea83c786d9d1835301d44877&stat=instructions:u

RKSimon commented:
https://llvm-compile-time-tracker.com/compare.php?from=ebc6c28a252915e6689ba7559bdc17eed0eea10c&to=9ced5e9548273789dc132ec033a31385ef429a21&stat=instructions:u

RKSimon commented:
any more comments?

RKSimon commented:
https://llvm-compile-time-tracker.com/compare.php?from=7c265e9f576041c20422218571bc2dacf20d65d9&to=bbe3c19fea298183e1d35e7f56f264c5ae386ad1&stat=instructions:u

RKSimon commented:
ping - anything else?

goldsteinn commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -3020,91 +3020,113 @@ SelectionDAG::getValidShiftAmountConstant(SDValue V,
+    // Use computeKnownBits to find a hidden constant (usually type legalized).
+    // e.g. Hidden behind multiple bitcasts/build_vector/casts etc.
+    KnownBits KnownAmt =
+        computeKnownBits(V.getOperand(1), DemandedElts, Depth + 1);
```
I don't think `Depth + 1` makes sense here. This isn't a recursive function. If its used by a recursive function, the caller should be setting `Depth + 1`.

goldsteinn commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -3020,91 +3020,113 @@ SelectionDAG::getValidShiftAmountConstant(SDValue V,
+      computeKnownBits(V.getOperand(1), DemandedElts, Depth + 1);
+  if (KnownAmt.getMaxValue().ult(BitWidth))
+    return KnownAmt.getMaxValue().getZExtValue();
+  return std::nullopt;
```
I think there should a single helper for all of the `getValid*ShiftAmount` functions.
Think it could just be `getValidShiftAmount` impl but return a `KnownBits` (the constant case we can just use `KnownBits::makeConstant()`).

Then `getValidShiftAmount` returns if the result is constant + less than bitwidth, `getValidMinimumShiftAmount` returns the getMinValue(), and vice versa for maximum. Think that will save a lot of dup code.

RKSimon replied:
Let me take a look, I'm wondering if ConstantRange will help us here. KnownBits alone tends to make the min/max values 'fuzzy'.

goldsteinn replied:
Its a bit of a mixed bag. KnownBits impl is more complete but has to throw away range information that isn't tied to particular bits. ConstantRange has a less complete impl, but obv represents what we want here better.

RKSimon committed:
[DAG] Add getValidShiftAmountRange to determine the range of valid shift amount values.

goldsteinn commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -3009,102 +3008,118 @@ SDValue SelectionDAG::getSplatValue(SDValue V, bool LegalTypes) {
+  // Use computeKnownBits to find a hidden constant/knownbits (usually type
+  // legalized). e.g. Hidden behind multiple bitcasts/build_vector/casts etc.
+  KnownBits KnownAmt =
+      computeKnownBits(V.getOperand(1), DemandedElts, Depth + 1);
```
Still think this should be `Depth` and the callers should be in charge of managing `Depth + 1` (the callers to `getValidShiftAmount`/`getValidMinimumShiftAmount`/`getValidMaximumShiftAmount` that is).

RKSimon replied:
I don't really mind either way - we do pass in the shift node and not the shift amount specifically so I can see both sides of it. This PR was just to pull out a minor tweak of #92096 and has instead turned into a monster :(

goldsteinn replied:
Its my preference, but not a req.

goldsteinn replied:
Guess my feeling is most of our recursive functions in DAGCombiner we put the `+1` at the initial callsite, so it might be confusing otherwise. Also this prohibits a 0 depth use.

RKSimon replied:
OK - the depth = 0 case could be an issue in the matchTruncateWithPACK use in X86ISelLowering.cpp - this is currently the only call from a non-recursive function, but there will probably be more in the future.

RKSimon commented:
https://llvm-compile-time-tracker.com/compare.php?from=f6c8e7dc3e1cbcecc2f01d898b895b96bb5723be&to=3879271a9f54e26d50948dc16b1b2ed0d0210431&stat=instructions:u

goldsteinn commented:
LGTM (with preference for changing `Depth` handling).

Wait a day or so before pushing so others have a chance to take a look.

RKSimon committed:
[DAG] getValidShiftAmountRange - move responsibility of the Depth operands to the caller

goldsteinn commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -2598,11 +2595,11 @@ bool TargetLowering::SimplifyDemandedBits(
-        if (!ShAmtC || ShAmtC->uge(BitWidth))
+        std::optional<uint64_t> ShAmtC =
+            TLO.DAG.getValidShiftAmount(Src, DemandedElts, Depth + 2);
+        if (!ShAmtC || *ShAmtC >= BitWidth)
```
Why Depth + 2? (here and above)

RKSimon replied:
These are 'inner' shifts of shifts. The originals should have used Depth+1 but that was missed.

RKSimon committed:
[DAG] getValidShiftAmountRange - fix upper bound of ConstantRange

I forgot to account for ConstantRange expecting the upper bound to be non-inclusive

RKSimon commented:
Thanks everyone, unless anyone has any more concerns I'm intending to commit this over the weekend.

RKSimon commented:
As usual when touching SimplifyDemandedBits et al this has managed to expose a bug somewhere else that only shows up in multi stage build bots - currently looking at this.
